containers/src/Data/IntMap/Internal.hs

```diff
@@ -761,6 +764,25 @@ disjoint t1@(Bin p1 m1 l1 r1) t2@(Bin p2 m2 l2 r2)
               | zero p1 m2       = disjoint t1 l2
               | otherwise        = disjoint t1 r2
 
+{--------------------------------------------------------------------
+  Compose
+--------------------------------------------------------------------}
+-- | /O(|ab|*log(|bc|))/. Relate the keys of one map to the values of
+-- the other, by using the values of the former as keys for lookups
+-- on the later.
+--
+-- > compose (fromList [('a', "A"), ('b', "B")]) (fromList [(1,'a'),(2,'b'),(3,'z')]) = fromList [(1,"A"),(2,"B")]
+--
+-- @
+-- ('compose' bc ab '!?') = (bc '!?') <=< (ab '!?')
+-- @
+--
+-- @since UNRELEASED
+compose :: IntMap c -> IntMap Int -> IntMap c
+compose bc ab = if null bc
+                then empty
```

Comment:
This is surprisingly lazy in the second argument. Is that special case a good special case? I'm not so convinced yet. Please make an argument.

Reply:
I don't think I yet sufficiently grasp the laziness expectations in these modules to answer properly. Here's what I managed to gather after reading your question:

The strict variants say
> If all values stored in all maps in the arguments are in WHNF, then all values stored in all maps in the results will be in WHNF once those maps are evaluated.

Does the same apply to keys for lazy maps given we have
>  A Map is strict in its keys but lazy in its values.

, and in that case `ab` should have its keys (and values, if strict) always evaluated if the result map is evaluated? Thus making optimizations like this one here somewhat break the interface?

Reply:
Oh I see, the second argument is potentially lazier than the first